### DISCUSSION TITLE
Waveshare RP2040-LoRa can't get GPS work

### DISCUSSION THREAD

**Discussion by radiohound** [meshtastic/firmware#5671] [Q&A] [answered]
Has anyone been successful in getting a GPS to work with the Waveshare RP2040-LoRa board? I have tried two different GPS's with no luck. I tried using 3 different pairs of Uart pins. I also tried connecting the same GPS's to a Heltec V3, and both GPS's worked the first try using that hardware.

Thanks for any help you can offer.

Walter

**Comment by fifieldt** [comment]
What do you see in the logs? https://flasher.meshtastic.org --> Open Serial Monitor ?

If you are building your own firmware, try setting #define GPS_DEBUG 1 in variant.h

**Comment by radiohound** [comment]
There's a lot going on, but I don't see any [GPS] references, like there are on devices that do connect to the gps. I am not seeing any serial errors, or anything. Is there something I should be looking for?

**Comment by fifieldt** [comment]
Give it a reboot. Early in the startup, there should be lines about detecting a GPS ... or not.

**Reply by radiohound** [reply]
Thank you. I have rebooted a number of times, but this device does not give me any GPS information in the serial debug output. When I try compiling myself, with the #define GPS_DEBUG 1, I get an error:

```
LittleFSDirImpl::openFile: rc=-2 fd=0x2000fe0c path=`/prefs/uiconfig.proto` openMode=0 accessMode=1 err=-2
ERROR | ??:??:?? 1 Could not open / read /prefs/uiconfig.proto
```

But it's my first time compiling Meshtastic, and a RP2040, so I am probably missing something.

**Reply by fifieldt** [reply]
Oh, you can safely ignore that error. Have you enabled GPS in settings in the app?

**Reply by radiohound** [reply]
Yes, I have been at this a while. I have tried many different firmware versions and set the gps up many times, in each one, and on separate serial pins. I even tried a Raspberry Pi Pico, to see if that would handle the gps initialization correctly. But that had no serial output about the GPS either.

**Reply by fifieldt** [reply]
Can you share your variant.h ?

**Reply by radiohound** [reply]
I am just using the stock one for RP2040-LoRa, with the GPS_DEBUG. However, I have also tried many of the precompiled versions too.

```
// #define RADIOLIB_CUSTOM_ARDUINO 1
// #define RADIOLIB_TONE_UNSUPPORTED 1
// #define RADIOLIB_SOFTWARE_SERIAL_UNSUPPORTED 1
#define GPS_DEBUG 1
#define ARDUINO_ARCH_AVR

// #define USE_SH1106 1

// default I2C pins:
// SDA = 4
// SCL = 5

// Recommended pins for SerialModule:
// txd = 8
// rxd = 9

#define EXT_NOTIFY_OUT 22
#define BUTTON_PIN -1 // Pin 17 used for antenna switching via DIO4

#define LED_PIN PIN_LED

// #define BATTERY_PIN 26
//  ratio of voltage divider = 3.0 (R17=200k, R18=100k)
// #define ADC_MULTIPLIER 3.1 // 3.0 + a bit for being optimistic

#define HAS_CPU_SHUTDOWN 1
#define USE_SX1262

#undef LORA_SCK
#undef LORA_MISO
#undef LORA_MOSI
#undef LORA_CS

// https://www.waveshare.com/rp2040-lora.htm
// https://www.waveshare.com/img/devkit/RP2040-LoRa-HF/RP2040-LoRa-HF-details-11.jpg
#define LORA_SCK 14  // GPIO14
#define LORA_MISO 24 // GPIO24
#define LORA_MOSI 15 // GPIO15
#define LORA_CS 13   // GPIO13

#define LORA_DIO0 RADIOLIB_NC // No GPIO connection
#define LORA_RESET 23         // GPIO23
#define LORA_BUSY 18          // GPIO18
#define LORA_DIO1 16          // GPIO16
#define LORA_DIO2 RADIOLIB_NC // Antenna switching, no GPIO connection
#define LORA_DIO3 RADIOLIB_NC // No GPIO connection
#define LORA_DIO4 17          // GPIO17

// On rp2040-lora board the antenna switch is wired and works with complementary-pin control logic.
// See PE4259 datasheet page 4

#ifdef USE_SX1262
#define SX126X_CS LORA_CS
#define SX126X_DIO1 LORA_DIO1
#define SX126X_BUSY LORA_BUSY
#define SX126X_RESET LORA_RESET
#define SX126X_DIO2_AS_RF_SWITCH // Antenna switch CTRL
#define SX126X_RXEN LORA_DIO4    // Antenna switch !CTRL via GPIO17
// #define SX126X_DIO3_TCXO_VOLTAGE 1.8
#endif
```

**Reply by fifieldt** [reply, marked as answer]
ok, try adding

#define HAS_GPS 1
#define GPS_TX_PIN (XXX) // This is for bits going TOWARDS the CPU - replace XXX with the GPIO number
#define GPS_RX_PIN (XXX) // This is for bits going TOWARDS the GPS - replace XXX with the GPIO number
#define PIN_SERIAL1_RX GPS_TX_PIN
#define PIN_SERIAL1_TX GPS_RX_PIN

**Reply by radiohound** [reply]
That gets it to look for the GPS at all different baud rates. However, I tried it with both GPS's, and it can't communicate with either one. 

```
DEBUG | ??:??:?? 41 [GPS] Probe for GPS at 38400
DEBUG | ??:??:?? 42 [GPS] Trying $PDTINFO (UC6580)...
DEBUG | ??:??:?? 42 [GPS] Trying $PDTINFO (UM600)...
DEBUG | ??:??:?? 43 [GPS] Trying $PCAS06,1*1A (ATGM336H)...
DEBUG | ??:??:?? 43 [GPS] Trying $PCAS06,1*1A (ATGM332D)...
DEBUG | ??:??:?? 44 [GPS] Trying $PAIR021*39 (AG3335)...
DEBUG | ??:??:?? 44 [GPS] Trying $PAIR021*39 (AG3352)...
DEBUG | ??:??:?? 45 [GPS] Trying $PQTMVERNO*58 (LC86)...
DEBUG | ??:??:?? 45 [GPS] Trying $PCAS06,0*1B (L76K)...
DEBUG | ??:??:?? 46 [GPS] Trying $PMTK605*31 (L76B)...
DEBUG | ??:??:?? 46 [GPS] Trying $PMTK605*31 (PA1616S)...
DEBUG | ??:??:?? 47 [GPS] 
WARN  | ??:??:?? 47 [GPS] No response for class 06 message 08
WARN  | ??:??:?? 47 [GPS] No GNSS Module (baudrate 38400)
DEBUG | ??:??:?? 47 [GPS] Probe for GPS at 4800
DEBUG | ??:??:?? 48 [GPS] Trying $PDTINFO (UC6580)...
DEBUG | ??:??:?? 48 [GPS] Trying $PDTINFO (UM600)...
DEBUG | ??:??:?? 49 [GPS] Trying $PCAS06,1*1A (ATGM336H)...
DEBUG | ??:??:?? 49 [GPS] Trying $PCAS06,1*1A (ATGM332D)...
DEBUG | ??:??:?? 50 [GPS] Trying $PAIR021*39 (AG3335)...
DEBUG | ??:??:?? 50 [GPS] Trying $PAIR021*39 (AG3352)...
DEBUG | ??:??:?? 51 [GPS] Trying $PQTMVERNO*58 (LC86)...
DEBUG | ??:??:?? 51 [GPS] Trying $PCAS06,0*1B (L76K)...
DEBUG | ??:??:?? 52 [GPS] Trying $PMTK605*31 (L76B)...
DEBUG | ??:??:?? 52 [GPS] Trying $PMTK605*31 (PA1616S)...
DEBUG | ??:??:?? 54 [GPS] 
WARN  | ??:??:?? 54 [GPS] No response for class 06 message 08
WARN  | ??:??:?? 54 [GPS] No GNSS Module (baudrate 4800)
```

**Reply by fifieldt** [reply]
Excellent, that's good progress.

So, with GPS DEBUG enabled, if there's anything at all from the GPS being transmitted you should see it in the log. Now check your TX and RX aren't the other way around, and the pin numbers are correct.

**Reply by radiohound** [reply]
Thank you so much! I went back to pins 0 and 1, and the gps's both worked there. I really appreciate your help on this! Without your extra code, it didn't work the first 25 times on those pins ...  :) 

Here is what worked for me.
```
#define HAS_GPS 1
#define GPS_TX_PIN (1) // This is for bits going TOWARDS the CPU - replace XXX with the GPIO number
#define GPS_RX_PIN (0) // This is for bits going TOWARDS the GPS - replace XXX with the GPIO number
#define PIN_SERIAL1_RX GPS_TX_PIN
#define PIN_SERIAL1_TX GPS_RX_PIN
```

**Reply by fifieldt** [reply]
Glad it's working!
